apis/core/lib/domain/queries/cube.ts

```diff
@@ -1,39 +1,55 @@
-import { CONSTRUCT } from '@tpluscode/sparql-builder'
+import { CONSTRUCT, DESCRIBE, SELECT } from '@tpluscode/sparql-builder'
 import StreamClient from 'sparql-http-client/StreamClient'
-import { cc, cube } from '@cube-creator/core/namespace'
-import { Stream, Term } from 'rdf-js'
-import { csvw, schema } from '@tpluscode/rdf-ns-builders'
+import * as ns from '@cube-creator/core/namespace'
+import { NamedNode, Stream, Term } from 'rdf-js'
+import { schema, sh } from '@tpluscode/rdf-ns-builders'
 import { Readable } from 'stream'
+import ParsingClient from 'sparql-http-client/ParsingClient'
+import $rdf from 'rdf-ext'
+import clownface from 'clownface'
 
-export function loadCubeShapes(dataset: Term, client: StreamClient): Promise<Stream & Readable> {
-  return CONSTRUCT`
-    ?s ?p ?o .
-    ?cube ${cube.observationConstraint} ?shape .
-    ?project ${cc.cubeGraph} ?cubeData .
-  `
+async function selectIdentifiers(datasetId: Term, parsingClient: ParsingClient): Promise<Record<string, NamedNode>> {
+  const [result] = await SELECT`?cube ?cubeData ?project ?shape`
     .WHERE`
-    GRAPH ?project {
-      ?project ${cc.dataset} ${dataset} ;
-               ${cc.cubeGraph} ?cubeData .
-    }
-
-    GRAPH ${dataset} {
-      ${dataset} ${schema.hasPart} ?cube
-    }
-
-    GRAPH ?cubeData {
-      ?s ?p ?o .
-      ?cube ${cube.observationConstraint} ?shape .
-
-      # Exclude non-observation resources mapped from CSV rows
-      MINUS { ?s ${csvw.describes} ?o }
-      MINUS { ?s ^${csvw.describes} ?row }
-
-      # Exclude cube details from result
-      MINUS { ?s a ${cube.Observation} }
-      MINUS { ?s a ${cube.Cube} }
-      MINUS { ?s a ${cube.ObservationSet} }
-      MINUS { ?s ${cube.observation} ?c }
-    }`
-    .execute(client.query)
+      GRAPH ?project {
+        ?project ${ns.cc.dataset} ${datasetId} ;
+                 ${ns.cc.cubeGraph} ?cubeData .
+      }
+
+      GRAPH ${datasetId} {
+        ${datasetId} ${schema.hasPart} ?cube
+      }
+
+      graph ?cubeData {
+        ?cube ${ns.cube.observationConstraint} ?shape .
+      }
+    `.execute(parsingClient.query)
+
+  return result as any
+}
+
+function constructPropertyShapes(shape: NamedNode, cubeData: NamedNode, streamClient: StreamClient) {
+  return CONSTRUCT`?s ?p ?o. ${shape} ?sp ?so .`
+    .FROM(cubeData)
+    .WHERE`
+      ${shape} ?sp ?so .
+      ${shape} ${sh.property} ?property .
+
+      ?property ${sh.path} ?path .
+      ?property (<>|!<>)* ?s .
+      ?s ?p ?o
```

Comment:
This is the core change. Rather than excluding triples from the result, the property shapes and their subgraphs are traversed

This is the crucial part is the restriction on `sh:path` property. Without it the query is two orders of magnitude slower